### ISSUE TITLE
Weight decay multiplier.

### ISSUE THREAD

**Issue by PigletOS** [leondgarse/Keras_insightface#5]
Hi, is there a way to set weight decay multiplier for the last layer in tensorflow or keras?
In the mxnet version, the author set the weight decay multiplier as 10 for the last dense layer.

**Comment by leondgarse** [comment]
I think we can, with something like this:
```py
class l2_decay_wdm(keras.regularizers.L2):
    def __init__(self, wd_func=None, **kwargs):
        super(l2_decay_wdm, self).__init__(**kwargs)
        self.wd_func = wd_func
    def __call__(self, x):
        self.l2 = self.wd_func()
        print("l2 =", self.l2)
        return super(l2_decay_wdm, self).__call__(x)
```
I will take a test, but not shure it will be useful..

**Comment by leondgarse** [comment]
I just make a checkin, and add a `output_wd_multiply` parameter to `train.Train`. [Main change is here](https://github.com/leondgarse/Keras_insightface/commit/c9b4b453859dcc5770eae3572c9f99728e86ca2c#diff-e44f4a60e89e820dde9bb27afa634965).
- If `output_wd_multiply` is set a value other than `1`, a `kernel_regularizer` will add to the `NormDense` layer. It is the `output_wd_multiply` value multiply with `self.optimizer.weight_decay`.
```py
kernel_regularizer = L2_decay_wdm(lambda: self.output_wd_multiply * self.optimizer.weight_decay)
```
- If comment out the print `tf.print(", l2 =", self.l2)` in `L2_decay_wdm`, it will print the `l2` value on every batch ends.
- I'm not shure if `output_wd_multiply` should still be `10` or not, as this `l2_regularizers` will work together with `optimizer.weight_decay`.

**Comment by PigletOS** [comment]
> I just make a checkin, and add a `output_wd_multiply` parameter to `train.Train`. [Main change is here](https://github.com/leondgarse/Keras_insightface/commit/c9b4b453859dcc5770eae3572c9f99728e86ca2c#diff-e44f4a60e89e820dde9bb27afa634965).
> 
> * If `output_wd_multiply` is set a value other than `1`, a `kernel_regularizer` will add to the `NormDense` layer. It is the `output_wd_multiply` value multiply with `self.optimizer.weight_decay`.
> 
> ```python
> kernel_regularizer = L2_decay_wdm(lambda: self.output_wd_multiply * self.optimizer.weight_decay)
> ```
> 
> * If comment out the print `tf.print(", l2 =", self.l2)` in `L2_decay_wdm`, it will print the `l2` value on every batch ends.
> * I'm not shure if `output_wd_multiply` should still be `10` or not, as this `l2_regularizers` will work together with `optimizer.weight_decay`.

Thanks！ That helps me a lot.  Now I am using weight decay on Mobilfacenet with 512 embedding size. Hope it work. ^V^

**Comment by leondgarse** [comment]
I dont think it's very effective..
This is my test using `mobilenet` + `casia dataset`
```py
import train, losses
import tensorflow_addons as tfa

data_basic_path = '/datasets/'
data_path = data_basic_path + 'faces_casia_112x112_folders'
eval_paths = [data_basic_path + ii for ii in ['faces_casia/lfw.bin', 'faces_casia/cfp_fp.bin', 'faces_casia/agedb_30.bin']]
basic_model = train.buildin_models("MobileNet", dropout=0, emb_shape=256)
tt = train.Train(data_path, save_path='T_adamw_5e5_dr0_soft_wdm20_E20_arc_casia_2.h5', eval_paths=eval_paths,
                 basic_model=basic_model, model=None, lr_base=0.001,
                 batch_size=64, random_status=3, output_wd_multiply=10)

optimizer = tfa.optimizers.AdamW(weight_decay=5e-5)
sch = [
    {"loss": keras.losses.CategoricalCrossentropy(label_smoothing=0.1), "epoch": 20, "optimizer": optimizer},
    {"loss": losses.ArcfaceLoss(), "bottleneckOnly": True, "epoch": 2},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.35, "epoch": 20},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.3, "epoch": 20},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.25, "epoch": 20},
    {"loss": losses.ArcfaceLoss(), "triplet": 64, "alpha": 0.2, "epoch": 20},
]

tt.train(sch, 0)
```

| output_wd_multiply | First 20 epochs | Second 80 Epochs |
| -------------------------- | -------------------- | ------------------------- |
| 0                              | Softmax            | Arcloss + triplet 64  |
| 0                              | Softmax            | Arcloss                    |
| 10                            | Softmax            | Arcloss + triplet 64  |
| 20                            | Softmax            | Arcloss + triplet 64  |

![mobilenet_casia_wdm](https://user-images.githubusercontent.com/5744524/93033657-c6786780-f669-11ea-9b6d-71c5ebfa636d.png)

**Comment by leondgarse** [comment]
I finnally figured out this. The problem is that the behavior of `weight_decay` in `mx.optimizer.SGD` and `tfa.optimizers.SGDW` is different.
- **MXNet SGD** multiplies `wd` with `lr`.
  ```py
  import mxnet as mx
  help(mx.optimizer.SGD)
  # rescaled_grad = lr * (rescale_grad * clip(grad, clip_gradient) + wd * weight)
  ```
  Test with `learning_rate=0.1, weight_decay=5e-4`, weight is actually modified by `5e-5`.
  ```py
  import mxnet as mx
  mm_loss_grad = mx.nd.array([[1., 1], [1, 1]])

  mm = mx.nd.array([[1., 1], [1, 1]])
  mopt = mx.optimizer.SGD(learning_rate=0.1)
  mopt.update(0, mm, mm_loss_grad, None)
  print(mm.asnumpy())  # Basic value is `mm - lr * mm_loss = 0.9`
  # [[0.9 0.9] [0.9 0.9]]

  mm = mx.nd.array([[1., 1], [1, 1]])
  mopt = mx.optimizer.SGD(learning_rate=0.1, wd=5e-4)
  mopt.update(0, mm, mm_loss_grad, None)
  print(mm.asnumpy())  # 0.9 - 0.89995 = 5e-5
  # [[0.89995 0.89995]  [0.89995 0.89995]]
  ```
- **tfa SGDW** behaves different, it does NOT multiply `wd` with `lr`. With `learning_rate=0.1, weight_decay=5e-4`, weight is actually modified with `5e-4`.
  ```py
  import tensorflow_addons as tfa
  ww_loss_grad = tf.convert_to_tensor([[1., 1.], [1., 1.]])
  ww = tf.Variable([[1., 1.], [1., 1.]])
  opt = tfa.optimizers.SGDW(learning_rate=0.1, weight_decay=5e-4)
  opt.apply_gradients(zip([ww_loss_grad], [ww]))
  print(ww.numpy()) # 0.9 - 0.8995 = 5e-4
  # [[0.8995 0.8995] [0.8995 0.8995]]
  ```
  So `learning_rate=0.1, weight_decay=5e-4` in `mx.optimizer.SGD` is equal to `learning_rate=0.1, weight_decay=5e-5` in `tfa.optimizers.SGDW`.
- Then on the `weight decay multiplier` question. If we set `wd_mult=10` in a MXNet layer, `wd` will mutiply by `10` in this layer.  [MXNet source code](https://github.com/apache/incubator-mxnet/blob/e6cea0d867329131fa6052e5f45dc5f626c00d72/python/mxnet/optimizer/optimizer.py#L482). This means it will be `weight_decay == 5e-4` in a keras layer.
- On the question `L2_regulalizer and weight_decay`, we can found some answers here [Weight Decay == L2 Regularization?](https://towardsdatascience.com/weight-decay-l2-regularization-90a9e17713cd) and here the arxiv paper [PDF DECOUPLED WEIGHT DECAY REGULARIZATION](https://arxiv.org/pdf/1711.05101.pdf).
  ```py
  ww_loss_grad = tf.convert_to_tensor([[1., 1.], [1., 1.]])
  ww = tf.Variable([[1., 1.], [1., 1.]])
  opt = keras.optimizers.SGD(0.1)
  with tf.GradientTape() as tape:
      # l2_loss = tf.nn.l2_loss(ww) * 5e-4
      l2_loss = keras.regularizers.L2(5e-4 / 2)(ww)  # `tf.nn.l2_loss` divided the loss by 2, `keras.regularizers.L2` not
  l2_grad = tape.gradient(l2_loss, ww).numpy()
  opt.apply_gradients(zip([ww_loss_grad + l2_grad], [ww]))
  print(ww.numpy()) # 0.9 - 0.89995 = 5e-5
  # [[0.89995 0.89995] [0.89995 0.89995]]
  ```
  That means the `L2_regulalizer` will modify the weights value by `l2 * lr == 5e-4 * 0.1 = 5e-5`.
- Finnaly, if we want the same result as `mx.optimizer.SGD(learning_rate=0.1, wd=5e-4)` and `wd_mult=10` in a MXNet layer, which actually decay this layer's weights with `wd * wd_mult * learning_rate == 5e-4`, and other layers `wd * learning_rate == 5e-5`.
  - Firstlly, the keras optimizer is `tfa.optimizers.SGDW(learning_rate=0.1, weight_decay=5e-5)`.
  - Then add a `keras.regularizers.L2` with `l2 == weight_decay / learning_rate * (wd_mult - 1) / 2` to this layer.
  ```py
  ww_loss_grad = tf.convert_to_tensor([[1., 1.], [1., 1.]])
  ww = tf.Variable([[1., 1.], [1., 1.]])
  opt = tfa.optimizers.SGDW(learning_rate=0.1, weight_decay=5e-5)
  with tf.GradientTape() as tape:
      l2_loss = keras.regularizers.L2(5e-5 / 0.1 * (10 - 1) / 2)(ww)
  l2_grad = tape.gradient(l2_loss, ww).numpy()
  opt.apply_gradients(zip([ww_loss_grad + l2_grad], [ww]))
  print(ww.numpy()) # 0.9 - 0.8995 = 5e-4
  # [[0.8995 0.8995] [0.8995 0.8995]]
  ```
- I'm still running a test on this and the code has been updated...

**Comment by leondgarse** [comment]
I would like to close this, as it's solved. [Detail explains on MXNet SGD weight_decay and tfa SGDW and l2 regularizer](https://github.com/leondgarse/Keras_insightface/discussions/19).
